content/issue-2/optimize-rust-comptime.md

```diff
@@ -0,0 +1,199 @@
+
+```
+tests/
+  foo.rs
+  bar.rs
+```
+
+Do this instead:
+
+```
+tests/
+  integration/
+    main.rs
+    foo.rs
+    bar.rs
+```
+
+It's because every file under `tests/` will be compiled into a separate binary (meaning every one will link dependencies). Apart from slow compilation, this can even slow down test runnings (a flaw in `cargo test`).
+
+This optimization didn't reduce our test time (probably due to the superiority of `cargo nextest`), but it immediately reduced the compilation time by another 2 minutes... It's also a bit funny that it also reduced the time for uploading/downloading, compressing/decompressing artifacts by 2 minutes...(although the latter did not affect the bottleneck).
+
+# Some previous efforts
+
+The above is the main process of my optimization this time, and now I can finally be satisfied with the work. Finally, I would like to summarize some of our previous efforts for reference.
+
+- Use [`cargo nextest`](https://github.com/nextest-rs/nextest) instead of `cargo test`.
+- Use the `workspace-hack` technique: see [`cargo hakari`](https://docs.rs/cargo-hakari/latest/cargo_hakari/about/index.html).
+- Add cache to the cargo registry, or use the recently stabilized [sparse index](https://blog.rust-lang.org/inside-rust/2023/01/30/cargo-sparse-protocol.html).
+- Split a huge crate into multiple smaller crates.
+- Try to reduce linking time: linking takes a lot of time and is single-threaded, so it may probably become a bottleneck.
+  - Use a faster linker: `mold` for Linux, `zld` for macOS. `lld` is the most mature option for production use.
+  - Turn off Link Time Optimization (LTO) on debug builds.
+- Trade-off between compile time and performance: The total time of CI is compile time + test time, so whether to turn on compile optimization (including LTO mentioned above), and how much to turn on, is actually a trade-off between the two. You can test and adjust that in order to achieve an overall optimal choice. For example, here's our build profile tuned by [BugenZhao](https://github.com/bugenzhao):
+
+```toml
+# The profile used for CI in pull requests.
+# External dependencies are built with optimization enabled, while crates in this workspace are built
+# with `dev` profile and full debug info. This is a trade-off between build time and e2e test time.
+[profile.ci-dev]
+inherits = "dev"
+incremental = false
+[profile.ci-dev.package."*"] # external dependencies
+opt-level = 1
+[profile.ci-dev.package."tokio"]
+opt-level = 3
+[profile.ci-dev.package."async_stack_trace"]
+opt-level = 3
+[profile.ci-dev.package."indextree"]
+opt-level = 3
+[profile.ci-dev.package."task_stats_alloc"]
+opt-level = 3
+
+# The profile used for deterministic simulation tests in CI.
+# The simulator can only run single-threaded, so optimization is required to make the running time
+# reasonable. The optimization level is customized to speed up the build.
+[profile.ci-sim]
+inherits = "dev"
+opt-level = 2
+incremental = false
+```
+
```

Comment:
I think we can also add the link: `Tips for Faster Rust Compile Times](https://endler.dev/2020/rust-compile-times/)` like the Chinese version.